Create expo-extension-migration guide (#21)
expo/fyi: expo-extension-migration.md

> # Migration off the expo extension

Reviewer: ```suggestion
# Migration off the expo fie extension
```

> > TL;DR: Starting in Expo SDK 40, the `.expo.js` extension is deprecated in favor of optional imports.

Reviewer: ```suggestion
> TL;DR: Starting in Expo SDK 41, the `.expo.js` extension is deprecated in favor of optional imports.
```

Author: It's removed in 41, deprecated in 40.

> Files with the `.expo.*` extension will no longer be used in SDK 41+, and the `EXPO_TARGET` environment variable is deprecated.

Reviewer: also the `--target` flag on publish/export/start

> Files with the `.expo.*` extension will no longer be used in SDK 41+, and the `EXPO_TARGET` environment variable is deprecated.

Author: ```suggestion
Files with the `.expo.*` extension will no longer be used in SDK 41+, the `EXPO_TARGET` environment variable, and the `--target` flag on `expo publish`, and `expo export` are all deprecated.
```

## Background

In the past, it was impossible to import files that didn't exist using the Metro bundler. Even if you wrapped a require statement in a try/catch, Metro would just fail to bundle instead of throwing. We've upstreamed a fix in SDK 40+ which correctly throws an error:

```ts
try {
  // If expo-camera doesn't exist, Metro would just crash instead of throwing :(
  require("expo-camera");
} catch (error) {
  // Starting in SDK 40+ the error will be thrown meaning you can optionally check if modules exist!
}
```

> To combat this, we created the `.expo.js` extension which would get used in the Expo client or legacy turtle V1 builds (`expo build`). This feature was primarily used by library authors who wanted to support standalone versions of expo packages like `react-native-camera` if `expo-camera` wasn't installed in the project. In SDK 40+ this can be done using try/catch syntax.

Reviewer: ```suggestion
To work around this, we created the `.expo.js` extension which would get used in the Expo client and classic standalone builds (`expo build`). This feature was primarily used by library authors who wanted to support alternative APIs like `react-native-camera` if `expo-camera` wasn't installed in the project. In SDK 40+ this can be done using try/catch syntax.
```

## Examples

### Alternative Modules

If you wanted to use `expo-camera` in your app when running in Expo Go, but some other package when running outside of Expo Go, you had to use a workflow extension.

#### Before

You used to have to create a "workflow specific" file extension before optional imports

`Camera.expo.js` (this file would be used in Expo environments)

```ts
export * from "expo-camera";
```

`Camera.js` (this file would be used in all other environments)

```ts
export * from "another-camera-package";
```

#### After

> Now you can simply use the optional import syntax:

Reviewer: ```suggestion
Now you can use the optional import syntax instead:
```

`Camera.js`

```js
let camera;
try {
  camera = require("expo-camera");
} catch {
  camera = require("another-camera-package");
}
module.exports = camera;
```

### Detecting Expo

> If you wanted to detect if you were running in Expo go, you needed to use workflow extensions, instead, you should use `expo-constants`.

Reviewer: ```suggestion
If you wanted to detect if you were running in Expo Go, you needed to use workflow extensions, instead, you should use `expo-constants`.
```

#### Before

`isExpo.expo.js` (this file would be used in Expo environments)

```ts
export const isExpo = true;
```

`isExpo.js` (this file would be used in all other environments)

```ts
export const isExpo = false;
```

#### After

Now you can use a far more reasonable approach:

```ts
let isExpo = false;
try {
  const Constants = require("expo-constants");

>   // True if the app is running in an `expo build` app or if it's running in Expo go.

Reviewer: ```suggestion
  // True if the app is running in an `expo build` app or if it's running in Expo Go.
```
